### DISCUSSION TITLE
Latency with ISend

### DISCUSSION THREAD

**Discussion by Deathn0t** [mpi4py/mpi4py#257] [Q&A] [answered]
Hello, 
When using `comm.Isend` I am having some unexpected latency (when calling `comm.Isend(...)` itself). Do you have any idea what could be the cause of this? As, this function should return "directly". I am using an `MPICH` backend for MPI. I can give more details of course if necessary. I am just interested to understand what is happening inside `comm.Isend(...)`.

Thank you for your help!

**Comment by dalcinl** [comment]
Don't get confused by naming. IIUC,  `MPI_Isend()` or any other non-blocking call should not necessarily return "directly" ir "immediately". It should return irrespective of what happens in other processes (i.e, irrespective of a matching `MPI_Recv()` being already posted). But it may still take some time to return, as the MPI progress engine can still take a chance of progressing other communication.

How much latency did you measure and how you did you measure it?
Does it change with small and large messages?
Does it change depending on whether the matching receive has been posted or not?

PS: Ultimately, you are asking in the wrong place. mpi4py does nothing special while wrapping the native MPI call. Well, not 100% true... mpi4py releases the GIL and that may have and impact on measured latency if other thread takes over. We should really ask MPICH folks for a definitive and authoritative answer. Maybe @roblatham00 or @hzhow can leave a quick note here.

**Comment by roblatham00** [comment, marked as answer]
I can only speak to the non-blocking I/O routines.   Sometimes those routines dispatch to the posix aio routines, but in quite a few cases the non-blocking MPI-IO routines do all the work when you post the operation.
